pyro/infer/predictive.py

```diff
@@ -0,0 +1,170 @@
+from functools import reduce
+import warnings
+
+import torch
+
+import pyro
+import pyro.poutine as poutine
+from pyro.poutine.util import prune_subsample_sites
+
+
+def _guess_max_plate_nesting(model, args, kwargs):
```

Comment:
I'm fine moving this helper if the logic is indeed duplicated. However beware falsely unifying subtly different functions. There was a refactoring effort a while ago to "deduplicate" logic among ELBO implementations, and the result was a morass of pyro/infer/enum.py that now breaks all dependency ordering and has no natural boundaries. I'd prefer three slightly different implementations than one weird conglomerate of things as disparate as, say probabilistic programming, deep reinforcement learning, and evolutionary algorithms.

Reply:
I think we can use the same implementation as the one in `elbo` and just pass the guide or an empty guide if no guide is provided. I will leave it to Du to make that determination. If it requires modifications to the utility function, then we should just replicate it as we are doing here.

Reply:
@fehiepsi : On second thoughts, this will require some changes inside HMC utilities, so its best to leave it as is right now to not complicate this PR.